packages/app/src/debug/utils/DebugMapping.ts

```diff
@@ -0,0 +1,68 @@
+import type { DebugSpecListSpecFragment, DebugSpecListTestsFragment, DebugLocalSpecsFragment, TestingTypeEnum, DebugSpecListGroupsFragment } from '../../generated/graphql'
+import { posixify } from '../../paths'
+
+type DebugSpecsArgs = {
+  specs: readonly DebugSpecListSpecFragment[]
+  tests: readonly DebugSpecListTestsFragment[]
+  groups: readonly DebugSpecListGroupsFragment[]
+  localSpecs: readonly DebugLocalSpecsFragment[]
+  currentTestingType: TestingTypeEnum
+}
+
+export type CloudDebugSpec = {
+  spec: DebugSpecListSpecFragment
+  tests: { [thumbprint: string]: DebugSpecListTestsFragment[] }
+  groups: { [groupId: string]: DebugSpecListGroupsFragment }
+  testingType: TestingTypeEnum
+  matchesCurrentTestingType: boolean
+  foundLocally: boolean
+}
+
+export const specsList = ({ specs, tests, localSpecs, currentTestingType, groups }: DebugSpecsArgs): CloudDebugSpec[] => {
+  const localSpecsSet = new Set(localSpecs.map(((spec) => posixify(spec.relative))))
+  const groupsMap = groups.reduce<{[id: string]: DebugSpecListGroupsFragment}>((acc, group) => ({ ...acc, [group.id]: group }), {})
+
+  const mappedTests = tests.reduce<{[id: string]: CloudDebugSpec}>((acc, curr) => {
+    let debugResult = acc[curr.specId]
+
+    if (!debugResult) {
+      const foundSpec = specs.find((spec) => spec.id === curr.specId)
+
+      if (!foundSpec) {
+        //TODO better handle error case
```

Comment:
Any concerns with this todo? Is there a place that this would be handled?

Reply:
@marktnoonan It would only happen if the Cloud gave back bad data.  I did just test it and it would cause the page to hang on the loading skeleton.  I will file a fast follow for this and update the comment.